src/EditorFeatures/Core/Implementation/TodoComment/TodoCommentIncrementalAnalyzer.cs

```diff
@@ -56,38 +57,46 @@ public async Task AnalyzeSyntaxAsync(Document document, InvocationReasons reason
+
+            // * NOTE * cancellation can't throw after this point.
+            if (existingData == null || existingData.Items.Length != newData.Items.Length)
+            {
+                Debug.Assert(_workspace == document.Project.Solution.Workspace);
+                RaiseTaskListUpdated(_workspace, document.Project.Solution, document.Id, newData.Items);
+            }
+        }
+
+        private static async Task<(Data existingData, Data newData)> ComputeAndPersistTodoCommentsAsync(
+            Document document,
+            TodoCommentState state,
+            TodoCommentTokens todoCommentTokens,
+            CancellationToken cancellationToken)
+        {
             // use tree version so that things like compiler option changes are considered
             var textVersion = await document.GetTextVersionAsync(cancellationToken).ConfigureAwait(false);
             var syntaxVersion = await document.GetSyntaxVersionAsync(cancellationToken).ConfigureAwait(false);
 
-            var existingData = await _state.TryGetExistingDataAsync(document, cancellationToken).ConfigureAwait(false);
+            var existingData = await state.TryGetExistingDataAsync(document, cancellationToken).ConfigureAwait(false);
             if (existingData != null)
             {
                 // check whether we can use the data as it is (can happen when re-using persisted data from previous VS session)
                 if (CheckVersions(document, textVersion, syntaxVersion, existingData))
                 {
-                    Debug.Assert(_workspace == document.Project.Solution.Workspace);
-                    RaiseTaskListUpdated(_workspace, document.Project.Solution, document.Id, existingData.Items);
-                    return;
+                    return (existingData, existingData);
                 }
             }
 
-            var tokens = _todoCommentTokens.GetTokens(document);
+            var tokens = todoCommentTokens.GetTokens(document);
             var comments = await GetTodoCommentsAsync(document, tokens, cancellationToken).ConfigureAwait(false);
             var items = await CreateItemsAsync(document, comments, cancellationToken).ConfigureAwait(false);
 
             var data = new Data(textVersion, syntaxVersion, items);
-            await _state.PersistAsync(document, data, cancellationToken).ConfigureAwait(false);
-
-            // * NOTE * cancellation can't throw after this point.
-            if (existingData == null || existingData.Items.Length > 0 || data.Items.Length > 0)
-            {
-                Debug.Assert(_workspace == document.Project.Solution.Workspace);
-                RaiseTaskListUpdated(_workspace, document.Project.Solution, document.Id, data.Items);
-            }
+            await state.PersistAsync(document, data, cancellationToken).ConfigureAwait(false);
```

Comment:
this line scares me.  is it guaranteed to be all or nothing?  i.e. if cancellation triggers could it ever be in an uncertain state?  should it not even take cancellation to be clear?

Reply:
Valid questions, but again this PR is not changing the original behavior around cancellation. I am just trying to fix the logic so that the API that tries to fetch task list items works correctly if the background analysis has never executed on the requested document.